src/Orleans/IDs/Interner.cs

```diff
@@ -96,31 +96,34 @@ public Interner(int initialSize, TimeSpan cleanupFreq)
         /// <param name="key">key to find</param>
         /// <param name="creatorFunc">function to create new object and store for this key if no cached copy exists</param>
         /// <returns>Object with specified key - either previous cached copy or newly created</returns>
-        public T FindOrCreate(K key, Func<T> creatorFunc)
+        public T FindOrCreate(K key, Func<K, T> creatorFunc)
         {
-            T obj = null;
-            WeakReference cacheEntry = internCache.GetOrAdd(key,
-                (k) =>
-                {
-                    obj = creatorFunc();
-                    return new WeakReference(obj);
-                });
-            if (cacheEntry != null)
+            T result;
+            WeakReference<T> cacheEntry;
+
+            // Attempt to get the existing value from cache.
+            internCache.TryGetValue(key, out cacheEntry);
+
+            // If no cache entry exists, create and insert a new one using the creator function.
+            if (cacheEntry == null)
             {
-                if (cacheEntry.IsAlive)
-                {
-                    // Re-use cached object
-                    obj = cacheEntry.Target as T;
-                }
+                result = creatorFunc(key);
+                cacheEntry = new WeakReference<T>(result);
+                internCache[key] = cacheEntry;
+                return result;
             }
-            if (obj == null)
+
+            // If a cache entry did exist, determine if it still holds a valid value.
+            cacheEntry.TryGetTarget(out result);
+            if (result == null)
             {
-                // Create new object
-                obj = creatorFunc();
-                cacheEntry = new WeakReference(obj);
-                obj = internCache.AddOrUpdate(key, cacheEntry, (k, w) => cacheEntry).Target as T;
+                // Create new object and ensure the entry is still valid by re-inserting it into the cache.
+                result = creatorFunc(key);
+                cacheEntry.SetTarget(result);
+                internCache[key] = cacheEntry;
```

Comment:
you are not worried about the different atomicity semantic snow (AddOrUpdate vs. what you do now)? 
I am not saying its wrong now, just drawing your attention that you changed it.